ffi/capi/src/locale.rs

```diff
@@ -21,21 +48,168 @@ pub type ICU4XLocale = Locale;
 /// terminator (this will just cause the function to panic, and is not a safety requirement).
 pub unsafe extern "C" fn icu4x_locale_create(value: *const u8, len: usize) -> *mut ICU4XLocale {
     let bytes = slice::from_raw_parts(value, len);
-    // todo: return errors
-    let loc = ICU4XLocale::from_bytes(bytes).unwrap();
-    Box::into_raw(Box::new(loc))
+    if let Ok(loc) = ICU4XLocale::from_bytes(bytes) {
+        Box::into_raw(Box::new(loc))
+    } else {
+        std::ptr::null_mut()
+    }
+}
+
+#[no_mangle]
+/// FFI version of [`Locale::clone()`]
+pub extern "C" fn icu4x_locale_clone(locale: &ICU4XLocale) -> *mut ICU4XLocale {
+    let clone = locale.clone();
+    Box::into_raw(Box::new(clone))
+}
+
+#[no_mangle]
+/// Write a string representation of the [`LanguageIdentifier`] part of
+/// [`ICU4XLocale`] to `write`.
+pub extern "C" fn icu4x_locale_basename(
+    locale: &ICU4XLocale,
+    write: &mut ICU4XWriteable,
+) -> ICU4XLocaleResult {
+    use writeable::Writeable;
+
+    let result = locale.id.write_to(write).is_ok();
+    write.flush();
+    if result {
+        ICU4XLocaleResult::Ok
+    } else {
+        ICU4XLocaleResult::Error
+    }
+}
+
+#[no_mangle]
+/// Write a string representation of the calendar extension to `write`
+pub extern "C" fn icu4x_locale_calendar(
```

Comment:
I feel like all of these should either be APIs on `Locale` that get exposed similarly to icu4x_locale_basename, or we should expose a higher level API for getting any subtag based on a string identifier.

I would like the FFI API to look _mostly_ like the Rust API, so if there are utility functions in FFI like this, they should be available in Rust too.

Reply:
I agree in principle, but since we're exposing opaque types through the C FFI, there's going to be cases where we'll want to have some accessor functions that we wouldn't need in Rust, because in Rust we can just work with the struct directly.

I've removed the special case unicode extension functions and replaced it with a generic `get_unicode_extension` function. I was on the fence about that when I initially wrote this code. I'd prefer to keep functions to access `language`, `script` and `region` directly rather than do some sort of string based dispatch for them.

Reply:
I see. I think having to convert a string to a key and then `.get()` is still somewhat tedious, but that's fair.

What do you think about the extension function wrapping one on Locale?

Reply:
Yes, now that I've thought about it some more, I agree we should not have things in the FFI that aren't present in the Rust API so we won't hit problems if we switch to code generation later on.